### DISCUSSION TITLE
Transfer Players Between Proxies

### DISCUSSION THREAD

**Discussion by awesomepandapig** [jeremylvln/Shulker#416] [Ideas]
In the most recent snapshot Mojang announced that Transfer packets will now be available for moving players between servers. https://www.minecraft.net/en-us/article/minecraft-snapshot-24w03a

I think it’d be really neat if during drain mode on the proxy, it could transfer players onto another proxy. This could even be integrated with Prometheus metrics to transfer players if a proxy is overloaded.

This would improve resource efficiency as proxies would not need to stay online (and costing $$$) with small amounts of players. They could simply be handed off and the proxy closed.

Since drain mode already excludes the proxy from the load balancer this could be as simple as transferring players to the load balancers external IP and which would then load balance them back onto another proxy.

**Comment by svetch** [comment]
This feature is already in development #364

**Reply by jeremylvln** [reply]
I had the exact same idea when they finally announced this packet!

As @Svetch said, I already did a WIP pull request. I will test once the new Minecraft version is released (not mentionning Paper).

I really like your idea of "merging" the players together to avoid having quite empty proxies. I never really dug the autoscaling part of Shulker but it's a really cool idea I would love to discuss more!

Let's hope Mojang releases the new version soon!!

**Comment by jeremylvln** [comment]
Hi @awesomepandapig!

This feature landed in Shulker 0.9.0. When the drain delay expires, the remaining players will be reconnected to the cluster (either using the LoadBalancer IP available in the associated `Service` or with a "preferred reconnection address" specified in the `ProxyFleet` (not yet in the documentation, it's under large revamping).

This adds a protection to avoid having players disconnected for regular maintenance events, I'm really excited Mojang added this packet!
